## 1. Import thư viện và load dữ liệu

In [1]:
# Import các thư viện cần thiết
import pandas as pd
import numpy as np

# Cấu hình hiển thị dữ liệu
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:,.2f}".format)

# Load dữ liệu
file_path = "vn_fuel_price_2018_present.csv"
df = pd.read_csv(file_path)

# Xem kích thước dữ liệu
print("Số dòng, số cột:", df.shape)

# Xem 5 dòng đầu tiên
display(df.head())

# Xem tên các cột
print("Danh sách cột:")
print(df.columns.tolist())

# Xem kiểu dữ liệu ban đầu
print("Kiểu dữ liệu ban đầu:")
display(df.dtypes)

Số dòng, số cột: (2946, 22)


,date,is_adjustment_day,usd_vnd,brent_usd_per_barrel,diesel_base_price,diesel_retail_price,diesel_bog_contribution,diesel_bog_spending,e5ron92_base_price,e5ron92_retail_price,e5ron92_bog_contribution,e5ron92_bog_spending,mazut_base_price,mazut_retail_price,mazut_bog_contribution,mazut_bog_spending,ron95_base_price,ron95_retail_price,ron95_bog_contribution,ron95_bog_spending,petrolimex_bog_ty_dong,pvoil_bog_ty_dong
0,08/05/2018,1,"22,805.00",74.16,"17,307.00","17,107.00",0.00,200.00,"20,398.00","19,440.00",0.00,958.00,"13,759.00","13,759.00",0.00,0.00,"21,362.00","21,362.00",0.00,0.00,"2,620.00",194.06
1,09/05/2018,0,"22,805.00",77.60,"17,307.00","17,107.00",0.00,200.00,"20,398.00","19,440.00",0.00,958.00,"13,759.00","13,759.00",0.00,0.00,"21,362.00","21,362.00",0.00,0.00,"2,620.00",194.06
2,10/05/2018,0,"22,805.00",77.59,"17,307.00","17,107.00",0.00,200.00,"20,398.00","19,440.00",0.00,958.00,"13,759.00","13,759.00",0.00,0.00,"21,362.00","21,362.00",0.00,0.00,"2,620.00",194.06
3,11/05/2018,0,"22,805.00",77.37,"17,307.00","17,107.00",0.00,200.00,"20,398.00","19,440.00",0.00,958.00,"13,759.00","13,759.00",0.00,0.00,"21,362.00","21,362.00",0.00,0.00,"2,620.00",194.06
4,12/05/2018,0,"22,805.00",77.64,"17,307.00","17,107.00",0.00,200.00,"20,398.00","19,440.00",0.00,958.00,"13,759.00","13,759.00",0.00,0.00,"21,362.00","21,362.00",0.00,0.00,"2,620.00",194.06


Danh sách cột:
['date', 'is_adjustment_day', 'usd_vnd', 'brent_usd_per_barrel', 'diesel_base_price', 'diesel_retail_price', 'diesel_bog_contribution', 'diesel_bog_spending', 'e5ron92_base_price', 'e5ron92_retail_price', 'e5ron92_bog_contribution', 'e5ron92_bog_spending', 'mazut_base_price', 'mazut_retail_price', 'mazut_bog_contribution', 'mazut_bog_spending', 'ron95_base_price', 'ron95_retail_price', 'ron95_bog_contribution', 'ron95_bog_spending', 'petrolimex_bog_ty_dong', 'pvoil_bog_ty_dong']
Kiểu dữ liệu ban đầu:


date                            str
is_adjustment_day             int64
usd_vnd                     float64
brent_usd_per_barrel        float64
diesel_base_price           float64
diesel_retail_price         float64
diesel_bog_contribution     float64
diesel_bog_spending         float64
e5ron92_base_price          float64
e5ron92_retail_price        float64
e5ron92_bog_contribution    float64
e5ron92_bog_spending        float64
mazut_base_price            float64
mazut_retail_price          float64
mazut_bog_contribution      float64
mazut_bog_spending          float64
ron95_base_price            float64
ron95_retail_price          float64
ron95_bog_contribution      float64
ron95_bog_spending          float64
petrolimex_bog_ty_dong      float64
pvoil_bog_ty_dong           float64
dtype: object

## 2. Kiểm tra giá trị thiếu, trùng lắp, ngoại lai, logic dữ liệu và chuẩn hóa dữ liệu


In [2]:
# Tạo bản sao để xử lý, tránh làm thay đổi dữ liệu gốc
df_clean = df.copy()

### 2.1 Kiểm tra giá trị thiếu

In [3]:
missing_summary = pd.DataFrame({
    "missing_count": df_clean.isna().sum(),
    "missing_percent": df_clean.isna().mean() * 100
}).sort_values(by="missing_count", ascending=False)

print("Bảng kiểm tra giá trị thiếu:")
display(missing_summary)

Bảng kiểm tra giá trị thiếu:


,missing_count,missing_percent
date,0,0.00
is_adjustment_day,0,0.00
usd_vnd,0,0.00
brent_usd_per_barrel,0,0.00
diesel_base_price,0,0.00
diesel_retail_price,0,0.00
diesel_bog_contribution,0,0.00
diesel_bog_spending,0,0.00
e5ron92_base_price,0,0.00
e5ron92_retail_price,0,0.00


Kết quả kiểm tra giá trị thiếu cho thấy tất cả các cột trong bộ dữ liệu đều có missing_count = 0 và missing_percent = 0.00%. Điều này nghĩa là dữ liệu không có giá trị bị thiếu ở các biến.

### 2.2 Kiểm tra dữ liệu trùng lắp

In [4]:
duplicate_count = df_clean.duplicated().sum()
print("Số dòng bị trùng lặp:", duplicate_count)

Số dòng bị trùng lặp: 0


### 2.3 Kiểm tra ngoại lai bằng phương pháp IQR

In [5]:
# Chọn các cột số để kiểm tra ngoại lai
numeric_cols = df_clean.select_dtypes(include=["int64", "float64"]).columns

outlier_summary = []

for col in numeric_cols:
    Q1 = df_clean[col].quantile(0.25)
    Q3 = df_clean[col].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outlier_count = ((df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)).sum()
    outlier_percent = outlier_count / len(df_clean) * 100
    
    outlier_summary.append({
        "column": col,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "lower_bound": lower_bound,
        "upper_bound": upper_bound,
        "outlier_count": outlier_count,
        "outlier_percent": outlier_percent,
        "outlier_percent_text": f"{outlier_percent:.2f}%"
    })

outlier_summary_df = pd.DataFrame(outlier_summary)
outlier_summary_df = outlier_summary_df.sort_values(by="outlier_count", ascending=False)

print("Bảng tổng hợp ngoại lai theo phương pháp IQR:")
display(outlier_summary_df)


Bảng tổng hợp ngoại lai theo phương pháp IQR:


,column,Q1,Q3,IQR,lower_bound,upper_bound,outlier_count,outlier_percent,outlier_percent_text
14,mazut_bog_spending,0.00,70.00,70.00,-105.00,175.00,512,17.38,17.38%
6,diesel_bog_spending,0.00,150.00,150.00,-225.00,375.00,403,13.68,13.68%
8,e5ron92_retail_price,"19,239.00","22,274.00","3,035.00","14,686.50","26,826.50",389,13.20,13.20%
15,ron95_base_price,"20,090.00","23,044.00","2,954.00","15,659.00","27,475.00",380,12.90,12.90%
12,mazut_retail_price,"14,163.25","16,524.00","2,360.75","10,622.12","20,065.12",298,10.12,10.12%
9,e5ron92_bog_contribution,0.00,100.00,100.00,-150.00,250.00,291,9.88,9.88%
11,mazut_base_price,"14,082.00","16,461.00","2,379.00","10,513.50","20,029.50",271,9.20,9.20%
7,e5ron92_base_price,"19,358.00","22,115.00","2,757.00","15,222.50","26,250.50",248,8.42,8.42%
0,is_adjustment_day,0.00,0.00,0.00,0.00,0.00,247,8.38,8.38%
17,ron95_bog_contribution,0.00,200.00,200.00,-300.00,500.00,221,7.50,7.50%


Kết quả kiểm tra ngoại lai bằng phương pháp IQR cho thấy một số biến có tỷ lệ ngoại lai khá cao. Trong đó, `mazut_bog_spending` có tỷ lệ ngoại lai cao nhất với 512 dòng, chiếm 17.38%; tiếp theo là `diesel_bog_spending` với 403 dòng, chiếm 13.68%. Một số biến giá xăng dầu như `e5ron92_retail_price`, `ron95_base_price`, `mazut_retail_price` cũng có tỷ lệ ngoại lai đáng chú ý, dao động từ khoảng 10% đến 13%.

Các ngoại lai này không được loại bỏ khỏi dữ liệu vì chúng có thể phản ánh biến động thực tế của thị trường xăng dầu, giá dầu thế giới, tỷ giá hoặc chính sách điều hành giá trong từng giai đoạn. Đặc biệt, các biến liên quan đến Quỹ bình ổn giá thường có nhiều giá trị bằng 0 và chỉ phát sinh mức trích lập/chi quỹ cao ở một số thời điểm, nên phương pháp IQR dễ đánh dấu các giai đoạn này là ngoại lai.

Biến `is_adjustment_day` cũng được phát hiện có 247 ngoại lai, chiếm 8.38%. Tuy nhiên, đây là biến nhị phân dùng để xác định ngày có điều chỉnh giá hay không, không phải biến số liên tục, nên không xem đây là ngoại lai cần xử lý.

Vì vậy, nhóm quyết định giữ nguyên các giá trị ngoại lai thay vì xóa bỏ. Việc giữ lại các giá trị này giúp bảo toàn các giai đoạn biến động quan trọng, phục vụ cho việc phân tích xu hướng, mức độ biến động và tác động của các yếu tố thị trường/chính sách đến giá xăng dầu tại Việt Nam.


### 2.4 Chuẩn hóa kiểu dữ liệu

In [6]:
# Chuyển cột date sang kiểu datetime
df_clean["date"] = pd.to_datetime(df_clean["date"], dayfirst=True, errors="coerce")

# Kiểm tra lại số dòng bị lỗi ngày tháng sau khi chuyển đổi
invalid_date_count = df_clean["date"].isna().sum()
print("Số dòng có ngày tháng không hợp lệ:", invalid_date_count)

# Chuyển is_adjustment_day sang kiểu boolean
df_clean["is_adjustment_day"] = df_clean["is_adjustment_day"].astype(bool)

# Chuyển các cột còn lại sang dạng số nếu cần
numeric_cols = df_clean.columns.drop(["date", "is_adjustment_day"])

for col in numeric_cols:
    df_clean[col] = pd.to_numeric(df_clean[col], errors="coerce")

# Sắp xếp dữ liệu theo thời gian
df_clean = df_clean.sort_values("date").reset_index(drop=True)

# Kiểm tra lại kiểu dữ liệu sau khi chuẩn hóa
print("Kiểu dữ liệu sau khi chuẩn hóa:")
display(df_clean.dtypes)

Số dòng có ngày tháng không hợp lệ: 0
Kiểu dữ liệu sau khi chuẩn hóa:


date                        datetime64[us]
is_adjustment_day                     bool
usd_vnd                            float64
brent_usd_per_barrel               float64
diesel_base_price                  float64
diesel_retail_price                float64
diesel_bog_contribution            float64
diesel_bog_spending                float64
e5ron92_base_price                 float64
e5ron92_retail_price               float64
e5ron92_bog_contribution           float64
e5ron92_bog_spending               float64
mazut_base_price                   float64
mazut_retail_price                 float64
mazut_bog_contribution             float64
mazut_bog_spending                 float64
ron95_base_price                   float64
ron95_retail_price                 float64
ron95_bog_contribution             float64
ron95_bog_spending                 float64
petrolimex_bog_ty_dong             float64
pvoil_bog_ty_dong                  float64
dtype: object

Sau khi chuẩn hóa, cột date đã được chuyển sang kiểu `datetime64[ns]` và không có dòng nào bị lỗi ngày tháng, với số dòng có ngày tháng không hợp lệ bằng 0. Điều này cho thấy dữ liệu thời gian hợp lệ và có thể sử dụng để phân tích xu hướng theo ngày, tháng, quý hoặc năm.

Bên cạnh đó biến `is_adjustment_day` được chuyển sang kiểu bool, phù hợp với ý nghĩa của biến là xác định ngày có điều chỉnh giá hay không.

### 2.5 Kiểm tra logic ở biến `is_adjustment_day`
Biến `is_adjustment`_day là biến nhị phân dùng để xác định một ngày có phải là ngày điều chỉnh giá xăng dầu hay không. Nên cần kiểm tra logic của biến này vì bộ dữ liệu được ghi nhận theo ngày, trong khi giá xăng dầu không thay đổi hằng ngày mà chỉ thay đổi vào các kỳ điều chỉnh. Nếu biến `is_adjustment_day` bị gán sai, các phân tích về số lần tăng/giảm giá, mức điều chỉnh giá và các giai đoạn biến động sẽ không chính xác.

In [7]:
# Các cột liên quan trực tiếp đến điều chỉnh giá trong nước
price_policy_cols = [
    "diesel_base_price",
    "diesel_retail_price",
    "diesel_bog_contribution",
    "diesel_bog_spending",
    "e5ron92_base_price",
    "e5ron92_retail_price",
    "e5ron92_bog_contribution",
    "e5ron92_bog_spending",
    "mazut_base_price",
    "mazut_retail_price",
    "mazut_bog_contribution",
    "mazut_bog_spending",
    "ron95_base_price",
    "ron95_retail_price",
    "ron95_bog_contribution",
    "ron95_bog_spending"
]

# Sắp xếp theo ngày
df_check = df_clean.sort_values("date").reset_index(drop=True)

# Kiểm tra xem ngày hiện tại có thay đổi giá/chính sách so với ngày trước đó không
df_check["has_price_policy_change"] = (
    df_check[price_policy_cols]
    .diff()
    .abs()
    .sum(axis=1) > 0
)

# Dòng đầu tiên không có ngày trước để so sánh, có thể xem là mốc bắt đầu
df_check.loc[0, "has_price_policy_change"] = df_check.loc[0, "is_adjustment_day"]

# So sánh biến gốc với biến tự tính
df_check["is_match"] = df_check["is_adjustment_day"] == df_check["has_price_policy_change"]

# Tổng kết kết quả kiểm tra
print("Tổng số dòng:", len(df_check))
print("Số dòng khớp:", df_check["is_match"].sum())
print("Số dòng không khớp:", (~df_check["is_match"]).sum())
print("Tỷ lệ khớp:", round(df_check["is_match"].mean() * 100, 2), "%")

Tổng số dòng: 2946
Số dòng khớp: 2946
Số dòng không khớp: 0
Tỷ lệ khớp: 100.0 %


Kiểm tra tính hợp lệ của biến `is_adjustment_day` bằng cách so sánh với sự thay đổi của các biến giá bán lẻ, giá cơ sở và các biến liên quan đến Quỹ bình ổn giá so với ngày liền trước. Kết quả cho thấy toàn bộ 2.946 dòng đều khớp, không có dòng nào không khớp, đạt tỷ lệ khớp 100%.

## 3. Khám phá ý nghĩa các bảng và biến

### 3.1 Tổng quan về bảng dữ liệu
Bộ dữ liệu `vn_fuel_price_2018_present.csv` ghi nhận theo tần suất ngày các thông tin liên quan đến giá xăng dầu bán lẻ tại Việt Nam, các yếu tố đầu vào (tỷ giá, giá dầu thô thế giới) và hoạt động của Quỹ bình ổn giá xăng dầu (Quỹ BOG), trải dài từ 08/05/2018 đến thời điểm hiện tại của dữ liệu. Mỗi dòng trong bảng tương ứng với một ngày, và mỗi cột là một biến số đo lường tại ngày đó.

In [8]:
# Xem lại danh sách cột và số lượng biến trong bảng
print("Số lượng biến (cột):", df_clean.shape[1])
print("Danh sách các biến:")
for i, col in enumerate(df_clean.columns, start=1):
    print(f"{i:>2}. {col}")

Số lượng biến (cột): 22
Danh sách các biến:
 1. date
 2. is_adjustment_day
 3. usd_vnd
 4. brent_usd_per_barrel
 5. diesel_base_price
 6. diesel_retail_price
 7. diesel_bog_contribution
 8. diesel_bog_spending
 9. e5ron92_base_price
10. e5ron92_retail_price
11. e5ron92_bog_contribution
12. e5ron92_bog_spending
13. mazut_base_price
14. mazut_retail_price
15. mazut_bog_contribution
16. mazut_bog_spending
17. ron95_base_price
18. ron95_retail_price
19. ron95_bog_contribution
20. ron95_bog_spending
21. petrolimex_bog_ty_dong
22. pvoil_bog_ty_dong


### 3.2 Diễn giải ý nghĩa từng biến
Bảng dưới liệt kê chi tiết ý nghĩa của toàn bộ 22 biến trong bộ dữ liệu:

| STT | Biến | Ý nghĩa | Kiểu dữ liệu |
|---|---|---|---|
| 1 | `date` | Ngày ghi nhận dữ liệu | datetime64 |
| 2 | `is_adjustment_day` | Cờ đánh dấu ngày đó có phải là kỳ điều chỉnh giá xăng dầu hay không | bool |
| 3 | `usd_vnd` | Tỷ giá quy đổi USD sang VND | float64 |
| 4 | `brent_usd_per_barrel` | Giá dầu thô Brent trên thị trường thế giới | float64 |
| 5 | `diesel_base_price` | Giá cơ sở của dầu Diesel | float64 |
| 6 | `diesel_retail_price` | Giá bán lẻ thực tế của dầu Diesel | float64 |
| 7 | `diesel_bog_contribution` | Mức trích lập Quỹ bình ổn giá (BOG) trên mỗi lít dầu Diesel bán ra | float64 |
| 8 | `diesel_bog_spending` | Mức chi Quỹ bình ổn giá (BOG) trên mỗi lít dầu Diesel bán ra | float64 |
| 9 | `e5ron92_base_price` | Giá cơ sở của xăng sinh học E5 RON 92 | float64 |
| 10 | `e5ron92_retail_price` | Giá bán lẻ thực tế của xăng E5 RON 92 | float64 |
| 11 | `e5ron92_bog_contribution` | Mức trích lập Quỹ bình ổn giá (BOG) trên mỗi lít xăng E5 RON 92 bán ra | float64 |
| 12 | `e5ron92_bog_spending` | Mức chiQuỹ bình ổn giá (BOG) trên mỗi lít xăng E5 RON 92 bán ra | float64 |
| 13 | `mazut_base_price` | Giá cơ sở của dầu mazut (dầu FO, dùng cho công nghiệp/hàng hải) | float64 |
| 14 | `mazut_retail_price` | Giá bán lẻ thực tế của dầu mazut | float64 |
| 15 | `mazut_bog_contribution` | Mức trích lập Quỹ bình ổn giá (BOG) trên mỗi kg dầu mazut bán ra | float64 |
| 16 | `mazut_bog_spending` | Mức chi Quỹ bình ổn giá (BOG) trên mỗi kg dầu mazut bán ra | float64 |
| 17 | `ron95_base_price` | Giá cơ sở của xăng RON 95 | float64 |
| 18 | `ron95_retail_price` | Giá bán lẻ thực tế của xăng RON 95 | float64 |
| 19 | `ron95_bog_contribution` | Mức trích lập Quỹ bình ổn giá (BOG) trên mỗi lít xăng RON 95 bán ra | float64 |
| 20 | `ron95_bog_spending` | Mức chi Quỹ bình ổn giá (BOG) trên mỗi lít xăng RON 95 bán ra | float64 |
| 21 | `petrolimex_bog_ty_dong` | Số dư Quỹ bình ổn giá xăng dầu của Tập đoàn Xăng dầu Việt Nam (Petrolimex) tại thời điểm ghi nhận | float64 |
| 22 | `pvoil_bog_ty_dong` | Số dư Quỹ bình ổn giá xăng dầu của Tổng công ty Dầu Việt Nam (PVOIL) tại thời điểm ghi nhận | float64 |

In [ ]:
# Kiểm chứng cách hiểu về mối quan hệ giữa 4 biến: base_price, bog_contribution, bog_spending, retail_price
products = ["diesel", "e5ron92", "mazut", "ron95"]

check_relation = pd.DataFrame({
    "product": products,
    "so_dong_khong_khop(base + contribution - spending = retail)": [
        (
            (df_clean[f"{p}_base_price"]
             + df_clean[f"{p}_bog_contribution"]
             - df_clean[f"{p}_bog_spending"]
             - df_clean[f"{p}_retail_price"]).abs() > 0.01
        ).sum()
        for p in products
    ]
})
display(check_relation)

,product,so_dong_khong_khop(base + contribution - spending = retail)
0,diesel,0
1,e5ron92,0
2,mazut,0
3,ron95,0


Kết quả đối chiếu công thức `giá bán lẻ = giá cơ sở + mức trích lập Quỹ BOG − mức chi Quỹ BOG` cho từng mặt hàng cho thấy cả 4 mặt hàng đều khớp trên toàn bộ 2.946 dòng, qua đó xác nhận đúng mối quan hệ giữa 4 biến giá trong cùng một nhóm mặt hàng. Về ý nghĩa nghiệp vụ, công thức này cũng hợp lý: khi Nhà nước trích lập vào Quỹ BOG, phần trích lập đó được cộng thêm vào giá bán lẻ (người mua trả thêm để góp vào quỹ); ngược lại khi Nhà nước chi Quỹ BOG để bình ổn giá, phần chi đó được trừ bớt khỏi giá bán lẻ (người mua được hỗ trợ giảm giá từ quỹ).

In [22]:
# Thống kê nhanh số kỳ điều chỉnh giá và khoảng cách trung bình giữa các kỳ điều chỉnh
adjustment_dates = df_clean.loc[df_clean["is_adjustment_day"], "date"].sort_values()
gap_days = adjustment_dates.diff().dt.days.dropna()

print("Tổng số kỳ điều chỉnh giá (is_adjustment_day = True):", df_clean["is_adjustment_day"].sum())
print("Khoảng cách trung bình giữa 2 kỳ điều chỉnh liên tiếp:", round(gap_days.mean(), 1), "ngày")
print("Khoảng cách nhỏ nhất / lớn nhất giữa 2 kỳ điều chỉnh:", gap_days.min(), "/", gap_days.max(), "ngày")

Tổng số kỳ điều chỉnh giá (is_adjustment_day = True): 247
Khoảng cách trung bình giữa 2 kỳ điều chỉnh liên tiếp: 12.0 ngày
Khoảng cách nhỏ nhất / lớn nhất giữa 2 kỳ điều chỉnh: 1.0 / 116.0 ngày


## 4. Kiểm tra cấu trúc dữ liệu

### 4.1 Kích thước và thông tin tổng quan bảng dữ liệu

In [11]:
# Kích thước bảng dữ liệu
print("Số dòng:", df_clean.shape[0])
print("Số cột:", df_clean.shape[1])
print()

# Thông tin tổng quan: tên cột, số lượng non-null, kiểu dữ liệu
df_clean.info()

Số dòng: 2946
Số cột: 22

<class 'pandas.DataFrame'>
RangeIndex: 2946 entries, 0 to 2945
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   date                      2946 non-null   datetime64[us]
 1   is_adjustment_day         2946 non-null   bool          
 2   usd_vnd                   2946 non-null   float64       
 3   brent_usd_per_barrel      2946 non-null   float64       
 4   diesel_base_price         2946 non-null   float64       
 5   diesel_retail_price       2946 non-null   float64       
 6   diesel_bog_contribution   2946 non-null   float64       
 7   diesel_bog_spending       2946 non-null   float64       
 8   e5ron92_base_price        2946 non-null   float64       
 9   e5ron92_retail_price      2946 non-null   float64       
 10  e5ron92_bog_contribution  2946 non-null   float64       
 11  e5ron92_bog_spending      2946 non-null   float64       
 12  mazut

### 4.2 Kiểu dữ liệu của từng biến

In [13]:
dtype_summary = pd.DataFrame({
    "dtype": df_clean.dtypes.astype(str),
    "so_gia_tri_khac_null": df_clean.notna().sum(),
    "so_gia_tri_duy_nhat": df_clean.nunique()
})
display(dtype_summary)

,dtype,so_gia_tri_khac_null,so_gia_tri_duy_nhat
date,datetime64[us],2946,2946
is_adjustment_day,bool,2946,2
usd_vnd,float64,2946,815
brent_usd_per_barrel,float64,2946,2492
diesel_base_price,float64,2946,239
diesel_retail_price,float64,2946,239
diesel_bog_contribution,float64,2946,21
diesel_bog_spending,float64,2946,18
e5ron92_base_price,float64,2946,243
e5ron92_retail_price,float64,2946,240


Bảng dữ liệu gồm 3 nhóm kiểu dữ liệu sau khi đã chuẩn hóa ở mục 2.4: biến thời gian (`date` — `datetime64[ns]`), biến nhị phân (`is_adjustment_day` — `bool`), và 20 biến số dạng `float64` còn lại (tỷ giá, giá dầu thô, giá xăng dầu, các biến Quỹ BOG).

### 4.3 Kiểm tra tính liên tục và tần suất của trục thời gian

In [14]:
# Khoảng thời gian của dữ liệu
date_min = df_clean["date"].min()
date_max = df_clean["date"].max()
n_days_expected = (date_max - date_min).days + 1

print("Ngày bắt đầu:", date_min.date())
print("Ngày kết thúc:", date_max.date())
print("Số ngày kỳ vọng nếu liên tục (không thiếu ngày nào):", n_days_expected)
print("Số dòng thực tế trong dữ liệu:", df_clean.shape[0])
print("Số ngày bị thiếu trong chuỗi thời gian:", n_days_expected - df_clean.shape[0])

if n_days_expected - df_clean.shape[0] > 0:
# Kiểm tra biến date có bị trùng lặp không (mỗi ngày chỉ nên xuất hiện 1 lần)
    print("\nSố ngày bị trùng lặp (xuất hiện nhiều hơn 1 dòng):", df_clean["date"].duplicated().sum())

    # Tìm cụ thể các ngày bị thiếu (nếu có)
    full_range = pd.date_range(start=date_min, end=date_max, freq="D")
    missing_dates = full_range.difference(df_clean["date"])
    print("Số lượng ngày bị thiếu cụ thể:", len(missing_dates))
    if len(missing_dates) > 0:
        print("Một số ngày bị thiếu:", list(missing_dates[:10].date))

Ngày bắt đầu: 2018-05-08
Ngày kết thúc: 2026-05-31
Số ngày kỳ vọng nếu liên tục (không thiếu ngày nào): 2946
Số dòng thực tế trong dữ liệu: 2946
Số ngày bị thiếu trong chuỗi thời gian: 0


Kết quả kiểm tra trục thời gian cho thấy dữ liệu được ghi nhận liên tục, mỗi ngày chỉ có đúng một dòng dữ liệu (không có ngày bị trùng lặp). Bảng dữ liệu đủ điều kiện để thực hiện các phân tích chuỗi thời gian mà không cần nội suy thêm cho các ngày bị thiếu.

### 4.4 Kiểm tra dải giá trị (range) và thống kê mô tả của các biến số

In [15]:
# Thống kê mô tả cho toàn bộ các biến số
numeric_cols = df_clean.select_dtypes(include=["float64"]).columns
desc_stats = df_clean[numeric_cols].describe().T
desc_stats["range"] = desc_stats["max"] - desc_stats["min"]
display(desc_stats)

,count,mean,std,min,25%,50%,75%,max,range
usd_vnd,"2,946.00","24,096.10","1,207.35","22,365.00","23,230.00","23,490.00","25,107.50","26,536.00","4,171.00"
brent_usd_per_barrel,"2,946.00",73.85,19.03,9.12,64.16,73.69,83.40,138.21,129.09
diesel_base_price,"2,946.00","18,576.05","4,045.07","9,857.00","16,586.00","18,408.00","20,278.50","42,841.00","32,984.00"
diesel_retail_price,"2,946.00","18,637.88","3,988.57","11,257.00","16,600.00","18,210.00","20,376.00","42,841.00","31,584.00"
diesel_bog_contribution,"2,946.00",186.84,284.28,0.00,0.00,0.00,300.00,"1,600.00","1,600.00"
diesel_bog_spending,"2,946.00",125.01,280.98,0.00,0.00,0.00,150.00,"1,500.00","1,500.00"
e5ron92_base_price,"2,946.00","20,703.11","3,110.61","10,942.00","19,358.00","20,470.00","22,115.00","31,302.00","20,360.00"
e5ron92_retail_price,"2,946.00","20,368.97","3,321.09","11,342.00","19,239.00","19,945.00","22,274.00","31,302.00","19,960.00"
e5ron92_bog_contribution,"2,946.00",72.16,150.30,0.00,0.00,0.00,100.00,950.00,950.00
e5ron92_bog_spending,"2,946.00",406.30,600.76,0.00,0.00,0.00,853.00,"2,000.00","2,000.00"


Kết quả kiểm tra dải giá trị cho thấy các biến giá cơ sở, giá bán lẻ, tỷ giá USD/VND và giá dầu thô Brent đều luôn dương trên toàn bộ 2.946 dòng, đúng với bản chất là các biến giá trị kinh tế. Các biến mức trích lập/chi Quỹ BOG theo lít/kg (`_bog_contribution`, `_bog_spending`) cũng không có giá trị âm, phù hợp với ý nghĩa là mức trích/chi cho mỗi đơn vị sản phẩm bán ra. Riêng hai biến số dư Quỹ BOG toàn doanh nghiệp (`petrolimex_bog_ty_dong`, `pvoil_bog_ty_dong`) có xuất hiện giá trị âm, phản ánh giai đoạn doanh nghiệp chi vượt mức trích lập (của Quỹ bình ổn giá).

Tổng kết: bảng dữ liệu có cấu trúc dạng chuỗi thời gian (time series) dài, đơn vị quan sát là ngày, với 2.946 dòng tương ứng 2.946 ngày liên tục không trùng lặp và không thiếu ngày, 22 cột với kiểu dữ liệu đã được chuẩn hóa phù hợp, và các biến số đều nằm trong dải giá trị hợp lý về mặt nghiệp vụ. Cấu trúc này đảm bảo dữ liệu sẵn sàng cho các bước phân tích thống kê mô tả, phân tích xu hướng và trực quan hóa theo thời gian ở các phần tiếp theo.